### DISCUSSION TITLE
Cant get RaspberryPi Pico spi1 working with touch...

### DISCUSSION THREAD

**Discussion by AJ6CU** [Bodmer/TFT_eSPI#2996] [Q&A - Touch]
I googled and read this forum extensively. Probably just need another set of eyes to figure out what I am doing wrong here.

I have a classic 2.8" display with ili9341 and I am trying to get it to work on a RaspberryPi Pico on spi1 and latest release of tft-espi (2.5.34) using the Arduino IDE 2.2.1.

What I have tried:
1. Display *is* working. I run the read/write example and pixles all come back passing. I load a complicated UX using lvgl and the UX display perfectly (same as on my Teensy 4 version of this board). My conclusion is the spi1 is working fine. Except for the chip selects, the SPI1 bus is bridged on the board to the Touch SPI.  (see photo of circuit below). 
2. Ran Read_User_sertup, and it reported the proper pins. 
3. I have confirmed continuity between Pico pins and the header the LCD plugs into. None are grounded either.
4. But when I try any of the touch examples, (calibrate, etc.), I get no touch events. 
5. Swapped both screens and processors, No change. Reflowed the solder joints on processor and header, no change.
6. Tried going back releases of TFT_espi, no change.
7. Tried using a different pin for Touch pin, no change. (GPIO17)
8. Reduced the frequency of touch read and spi, no change.
9. Grasping at straws, defined the pins both as Dnn  and nn, no change.

Must be missing something obvious. Hopefully  a fresh set of eyes will spot it. Getting to the point where I will just redesign the board to use spi0...

The custom setup file:

`//                            USER DEFINED SETTINGS
//   Set driver type, fonts to be loaded, pins used and SPI control method etc
//
//   See the User_Setup_Select.h file if you wish to be able to define multiple
//   setups and then easily select which setup file is used by the compiler.
//
//   If this file is edited correctly then all the library example sketches should
//   run without the need to make any more changes for a particular hardware setup!
//   Note that some sketches are designed for a particular TFT pixel width/height

#define USER_SETUP_ID 560
// ##################################################################################
//
// Section 1. Call up the right driver file and any options for it
//
// ##################################################################################

// Tell the library to use 8 bit parallel mode (otherwise SPI is assumed)
//#define TFT_PARALLEL_8_BIT

// Display type -  only define if RPi display
//#define RPI_DISPLAY_TYPE // 20MHz maximum SPI

// Only define one driver, the other ones must be commented out
#define ILI9341_DRIVER
//#define ST7735_DRIVER      // Define additional parameters below for this display
//#define ILI9163_DRIVER     // Define additional parameters below for this display
//#define S6D02A1_DRIVER
//#define RPI_ILI9486_DRIVER // 20MHz maximum SPI
//#define HX8357D_DRIVER
//#define ILI9481_DRIVER
//#define ILI9486_DRIVER
//#define ILI9488_DRIVER     // WARNING: Do not connect ILI9488 display SDO to MISO if other devices share the SPI bus (TFT SDO does NOT tristate when CS is high)
//#define ST7789_DRIVER      // Full configuration option, define additional parameters below for this display
//#define ST7789_2_DRIVER    // Minimal configuration option, define additional parameters below for this display
//#define R61581_DRIVER
//#define RM68140_DRIVER
//#define ST7796_DRIVER
//#define SSD1963_480_DRIVER
//#define SSD1963_800_DRIVER
//#define SSD1963_800ALT_DRIVER
//#define ILI9225_DRIVER

// Some displays support SPI reads via the MISO pin, other displays have a single
// bi-directional SDA pin and the library will try to read this via the MOSI line.
// To use the SDA line for reading data from the TFT uncomment the following line:

// #define TFT_SDA_READ      // This option is for ESP32 ONLY, tested with ST7789 display only

// For ST7735, ST7789 and ILI9341 ONLY, define the colour order IF the blue and red are swapped on your display
// Try ONE option at a time to find the correct colour order for your display

//  #define TFT_RGB_ORDER TFT_RGB  // Colour order Red-Green-Blue
//  #define TFT_RGB_ORDER TFT_BGR  // Colour order Blue-Green-Red

// For ST7789, ST7735 and ILI9163 ONLY, define the pixel width and height in portrait orientation
// #define TFT_WIDTH  80
// #define TFT_WIDTH  128
// #define TFT_WIDTH  240 // ST7789 240 x 240 and 240 x 320
// #define TFT_HEIGHT 160
// #define TFT_HEIGHT 128
// #define TFT_HEIGHT 240 // ST7789 240 x 240
// #define TFT_HEIGHT 320 // ST7789 240 x 320

// For ST7735 ONLY, define the type of display, originally this was based on the
// colour of the tab on the screen protector film but this is not always true, so try
// out the different options below if the screen does not display graphics correctly,
// e.g. colours wrong, mirror images, or tray pixels at the edges.
// Comment out ALL BUT ONE of these options for a ST7735 display driver, save this
// this User_Setup file, then rebuild and upload the sketch to the board again:

// #define ST7735_INITB
// #define ST7735_GREENTAB
// #define ST7735_GREENTAB2
// #define ST7735_GREENTAB3
// #define ST7735_GREENTAB128    // For 128 x 128 display
// #define ST7735_GREENTAB160x80 // For 160 x 80 display (BGR, inverted, 26 offset)
// #define ST7735_REDTAB
// #define ST7735_BLACKTAB
// #define ST7735_REDTAB160x80   // For 160 x 80 display with 24 pixel offset

// If colours are inverted (white shows as black) then uncomment one of the next
// 2 lines try both options, one of the options should correct the inversion.

//#define TFT_INVERSION_ON
//#define TFT_INVERSION_OFF

// ##################################################################################
//
// Section 2. Define the pins that are used to interface with the display here
//
// ##################################################################################

// If a backlight control signal is available then define the TFT_BL pin in Section 2
// below. The backlight will be turned ON when tft.begin() is called, but the library
// needs to know if the LEDs are ON with the pin HIGH or LOW. If the LEDs are to be
// driven with a PWM signal or turned OFF/ON then this must be handled by the user
// sketch. e.g. with digitalWrite(TFT_BL, LOW);

// #define TFT_BL   32            // LED back-light control pin
//#define TFT_BACKLIGHT_ON HIGH  // Level to turn ON back-light (HIGH or LOW)

// We must use hardware SPI, a minimum of 3 GPIO pins is needed.
// Typical setup for the RP2040 is :
//
// Display SDO/MISO  to RP2040 pin D0 (or leave disconnected if not reading TFT)
// Display LED       to RP2040 pin 3V3 or 5V
// Display SCK       to RP2040 pin D2
// Display SDI/MOSI  to RP2040 pin D3
// Display DC (RS/AO)to RP2040 pin D18 (can use another pin if desired)
// Display RESET     to RP2040 pin D19 (can use another pin if desired)
// Display CS        to RP2040 pin D20 (can use another pin if desired, or GND, see below)
// Display GND       to RP2040 pin GND (0V)
// Display VCC       to RP2040 5V or 3.3V (5v if display has a 5V to 3.3V regulator fitted)
//
// The DC (Data Command) pin may be labelled AO or RS (Register Select)
//
// With some displays such as the ILI9341 the TFT CS pin can be connected to GND if no more
// SPI devices (e.g. an SD Card) are connected, in this case comment out the #define TFT_CS
// line below so it is NOT defined. Other displays such at the ST7735 require the TFT CS pin
// to be toggled during setup, so in these cases the TFT_CS line must be defined and connected.

// For the Pico use these #define lines

#define TFT_MISO  D12
#define TFT_MOSI  D11
#define TFT_SCLK  D10
#define TFT_CS   D13  // Chip select control pin
#define TFT_DC   D8  // Data Command control pin
//#define TFT_RST  255  // Reset pin (could connect to Arduino RESET pin)
//#define TFT_BL     // LED back-light

#define TOUCH_CS D9     // Chip select pin (T_CS) of touch screen

//#define TFT_MISO  D12
//#define TFT_MOSI  D11
//#define TFT_SCLK  D10
//#define TFT_CS   D13  // Chip select control pin
//#define TFT_DC   D8  // Data Command control pin
//#define TFT_RST  255  // Reset pin (could connect to Arduino RESET pin)
//#define TFT_BL     // LED back-light

//#define TOUCH_CS D9     // Chip select pin (T_CS) of touch screen

// ##################################################################################
//
// Section 3. Define the fonts that are to be used here
//
// ##################################################################################

// Comment out the #defines below with // to stop that font being loaded
// The ESP8366 and ESP32 have plenty of memory so commenting out fonts is not
// normally necessary. If all fonts are loaded the extra FLASH space required is
// about 17Kbytes. To save FLASH space only enable the fonts you need!

#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:-.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
//#define LOAD_FONT8N // Font 8. Alternative to Font 8 above, slightly narrower, so 3 digits fit a 160 pixel TFT
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts

// Comment out the #define below to stop the SPIFFS filing system and smooth font code being loaded
// this will save ~20kbytes of FLASH
#define SMOOTH_FONT

// ##################################################################################
//
// Section 4. Other options
//
// ##################################################################################

// For the RP2040 processor define the SPI port channel used, default is 0
#define TFT_SPI_PORT 1 // Set to 0 if SPI0 pins are used, or 1 if spi1 pins used

// Define the SPI clock frequency, this affects the graphics rendering speed. Too
// fast and the TFT driver will not keep up and display corruption appears.
// With an ILI9341 display 40MHz works OK, 80MHz sometimes fails
// With a ST7735 display more than 27MHz may not work (spurious pixels and lines)
// With an ILI9163 display 27 MHz works OK.

// #define SPI_FREQUENCY   1000000
// #define SPI_FREQUENCY   5000000
// #define SPI_FREQUENCY  10000000
// #define SPI_FREQUENCY  20000000
//#define SPI_FREQUENCY  32000000
#define SPI_FREQUENCY  70000000

// Optional reduced SPI frequency for reading TFT
#define SPI_READ_FREQUENCY  20000000

// The XPT2046 requires a lower SPI clock rate of 2.5MHz so we define that here:
#define SPI_TOUCH_FREQUENCY  2500000

`
The kicad circuit is:

![image](https://github.com/Bodmer/TFT_eSPI/assets/70183884/d82e9ed1-9293-44fd-be4d-67421691777e)

Hope someone can point out my error!

thanks
Mark

**Comment by AJ6CU** [comment]
Becoming convinced that this is a hardware problem with perhaps a little frequency read speed mixed in. Put the 2.8 screen on the end of a 25cm ribbon cable and the test touch example is showing values (although some seem to be noisy) that are reactive to position and pressure.  The calibrate test self completes suggesting recognizing noise as a press.  Strangely, the test touch example still seems to work when the cable is bent so the screen is back at its original position, (just with a 25cm cable inbetween). 

Suggestion would be appreciated, but at this point very unlikely a software issue.

Mark

**Reply by TimH615** [reply]
I have exactly the same problem.  I've been using ILI9341's and TFT_eSPI extensively with SAMD21's without any issues.  I'm pretty sure a few months ago I had the example 320x240 keypad working on a pico.  But now when it tries to do the calibration, it runs through the 4 corners without being pressed (noise?).  I'm using dupont connectors so the wire runs are about 7 inches between the display and the pico.  The graphics work fine.  When I try running other XPT2046 libraries I also have problems.  Also I actually have six ILI9341 boards and swapping out the boards has no effect so it's not a defective display.  I've also tried going back to old TFT_eSPI versions -- 2.5 has the same problem and 2.4 didn't compile for pico (IIRC).  Did you solve this problem?  Thanks for any help.  Also, my problem (and yours?) is the same one in this post from 2022:
https://forum.arduino.cc/t/ili9341-tft-espi-touch-calibrate-example-finishes-without-touching-the-screen/1004912

**Reply by AJ6CU** [reply]
Unfortunately, I have not resolved this one either (using SPI1 with port set appropriately in the config file). The only thing that seems to provide any touch response is when I extend the display via the 25cm ribbon cable. But still a lot of noise when I run the touch test example. But I do get different x,y, and z values with the cable and I touch the screen. The need to introduce this cable suggests either a noise issue or a timing one (same code, just cable length).

Funny thing is I "know" I had this working on a prior incarnation of the board and the "only" ;-) change here is that I reposition the pins on the board to match the lcd to avoid crossovers and splits (for the spi bus).

I remember coming across a library mod that needed to be made for the spi1 and pico. This was *different* than the port option provided in the 2040 config. And of course I didn't save this mod and it gog smashed when I uploaded....  Tried going thru my browser histroy but could not find it...  For a while, I thought that the SPI_Mode (controls whether the detection happens on rising/falling etc. edge) was the magic. But went thru all 4 values (0-3) with no luck.

A little tired of knocking my head against the wall using SPI1. Ater the holidays, going to respin the board to use SPI0. Will also pay closer attention to the size and location of the traces for the SPI bus to the connector. Perhaps I am introducing noise in the bus.

On a related topic, Bodmer observed that the Arduino IDE package maintainer for the Pico ("Earle") had updated his library to use PXXX to maintain compatibility with the mbed board package. Might give that a try before I redesign the board, but I think this is more likely a "hail Mary" attempt.

Let me know if you solve this too!

Regards
Mark

**Comment by larrybud2004** [comment]
I have exactly the same problem. Can't get touch to work on spi0 or spi1. Bought a 2nd screen (HiLetGo variety) thinking it might be hardware. No go. Also tried a couple of different Picos.

Additional info, I get whacky data for pressure, which triggers a false positive touch:

15:07:55.270 -> x: 0     y: 7936     z: 64767 
15:07:55.485 -> x: 0     y: 7936     z: 62975 
15:07:55.777 -> x: 0     y: 7936     z: 2559 
15:07:56.011 -> x: 0     y: 7936     z: 1535 
15:07:56.271 -> x: 0     y: 7936     z: 2559 
15:07:56.486 -> x: 0     y: 7936     z: 62975 
15:07:56.764 -> x: 0     y: 7936     z: 511 
15:07:57.027 -> x: 0     y: 7936     z: 3071 
15:07:57.264 -> x: 0     y: 7936     z: 2047 
15:07:57.527 -> x: 0     y: 7936     z: 63743 
15:07:57.759 -> x: 0     y: 7936     z: 3071 
15:07:58.023 -> x: 0     y: 7936     z: 2815 
15:07:58.256 -> x: 0     y: 7936     z: 62975 
15:07:58.520 -> x: 0     y: 7936     z: 63999 
15:07:58.782 -> x: 0     y: 7936     z: 3583 
15:07:59.016 -> x: 0     y: 7936     z: 62975

**Reply by AJ6CU** [reply]
I tried breadboarding the screen to spi0. Also got lots of what looks like noise without pressure. With pressure, the numbers were more consistent, but still many outliers around the "proper" value. Can't remember if my Z varied as much as yours, but certainly had more variations in my x and y. But then, couldn't get display to work. I set it aside thinking that I must of misswired something in the breadboard. 

Regards,

Mark

**Comment by TimH615** [comment]
Well it's gratifying that it isn't just me doing something stupid or forgetting to do something.  Gonna wire up a Teensy 4.0, which cost as much as six picos, to a 9341 display tonight to reassure myself that it's a pico issue.   I've tried different XPT2046 libraries with the pico and they don't seem to work either.   So it must be some signal interference issue.  I'm building touchscreen-LoRa radios and have had other RF issues which I fixed by putting devices in paper sleeves lined on the outside with copper tape.  Will try that.  Incidentially Mark I'm a ham-- N1PAZ.  Am going to use the LoRa setup to experiment with radio beacons for navigation. -- Tim

**Reply by AJ6CU** [reply]
Tim,

I had no problem with the Teensy 4.0 (Haven't tried it on the 4.1 yet). Before re-spinning my Pico board, I am going to do some googling on SPI noise reduction and see if there should be some caps somewhere along the bus (or not). It is indeed strange that the direct socket on the board is no-go for touch (but video works fine). And to get anything going with touch, I needed to added a ribbob cable to create some distance. My reading is that on the SPI bus, shorter is better...

LoRa radios sounds like an interesting application!.  My use is to rewrite the KD8CEC firmware for the uBITX (from HF Signals) that was written fo rNextion displays to use standard tft screens with touch. With the power of a pico, (and perhaps using that dual processor to provide a web page for control), Been using lvgl.io on top of tft_espi. The SquareLine GUI builder for lvgl really helps to build the UX quickly!

73
Mark

**Comment by Bodmer** [comment]
See #3009

**Reply by AJ6CU** [reply]
The suggestion did restored touch on spi1 but I needed to put the display at the end of a 25cm ribbon cable from the processor board.  Directly pluggied into the processor board I get no touch response. 

The display of the user interface works fine, just no touch unless I add that ribbon cable inbetween. BTW: I did try slowing down the SPI_Touch_Frequency to 1500000 and the SPI_Frequency to 32000000.

Any other software possibilities or is this likely a noise issue on the spi1 bus. 

Regards
Mark

**Reply by Bodmer** [reply]
I suspect this is a crosstalk issue in the ribbon cable. Sperating out the ribbon wires or making every other wire a 0v/GND may help.

e.g instead of the wires being:
signal
signal
signal
.... etc

Use:
signal
GND
signal
GND
signal

**Comment by AJ6CU** [comment]
Thank you for the suggestion! 

The strange thing is that ribbon cable *works* and if I plug the 2.8 tfrt into the socket on the back of the processor board (it matches the tft pinout), it does not work....  And for a similar designed board for the Teensy, it works fine (of course different  pins).   

I suspect that your advice applies to how I designed the traces that connect the tft socket to their respective pins... Probably need to look specifically at the touch select and the MISO pin routings....

Regards

Mark

**Comment by ballistRC** [comment]
I guess this is solved but if not make sure TOUCH_CS is setup in the User_Setup.h located in the lib directory.  Check the build log that it is really seen.

**Comment by AJ6CU** [comment]
AnthonyThanks for the suggestions!  Unfortunately I did all these. The only way I could get any response was to separate the screen from the processor board by a 25cm ribbon cable. Then things seemed to work. At this point I am thinking that I have noise on the lines on the processor board on the data in mode. Doesn’t look like a software error (either in eSPI or my code). Would be excited to hear that someone else got eSPI working on SPI1 and a Pico. That would make it 100% my hardware problem. RegardsMarkSent from my iPhoneOn Dec 19, 2023, at 2:32 AM, Anthony Clark ***@***.***> wrote:﻿
I guess this is solved but if not make sure TOUCH_CS is setup in the User_Setup.h located in the lib directory.  Check the build log that it is really seen.
